### PR TITLE
NodeJS 4.3

### PR DESCRIPTION
**Pull request by garymacindoe**
AWS have deprecated and removed support for lambdas written in NodeJS 0.10.  They have also changed the signature of the Lambda handler function from `handler(event, context)` to `handler(event, context, callback)`.  The `callback` argument replaces the `context.fail`, `context.succeed` and `context.done` methods (although if you log the `context` object in AWS you will see the `context.done` method still exists and is probably called by `callback` seeing as the behaviour is the same).

This PR adds support for lambdas written in NodeJS 4.3+ (including 6.10) without removing the existing support for lambdas written in NodeJS 0.10.  The tests have been updated to work with NodeJS 4.3 but several packages have had to be upgraded in the process and as a result the tests no longer work with NodeJS 0.10.

### CODE REVIEW

**Comment by rvedotrc** [comment]
I note that you've dropped support for context.fail and context.succeed (i.e. we do not provide them, and Lambda does); whereas context.done remains. But our implementation will never be perfect – even context.done and the last-arg callback are meant to do different things, but we don't follow that. See http://docs.aws.amazon.com/lambda/latest/dg/nodejs-prog-model-context.html#nodejs-prog-model-context-properties "callbackWaitsForEmptyEventLoop".

In summary: not perfect, but we will never achieve perfection. This is still a helpful PR. Good enough ;-)
